src/components/Editable.js

```diff
@@ -188,15 +166,65 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
   }
 
   // using useRef hook to store throttled function so that it can persist even between component re-renders, so that throttle.flush method can be used properly
-  const throttledChangeRef = useRef(_.throttle(onChangeHandler, EDIT_THROTTLE, { leading: false }))
+  const throttledChangeRef = useRef(throttle(thoughtChangeHandler, EDIT_THROTTLE))
+
   ShortcutEmitter.on('shortcut', () => {
     throttledChangeRef.current.flush()
   })
 
-  useEffect(() => throttledChangeRef.current.flush, []) // clean up function when component unmounts (flushing throttle change)
+  const onChangeHandler = e => {
+    // NOTE: When Subthought components are re-rendered on edit, change is called with identical old and new values (?) causing an infinite loop
+    const newValue = he.decode(strip(e.target.value))
+    const oldValue = oldValueRef.current
+
+    // TODO: Disable keypress
+    // e.preventDefault() does not work
+    // disabled={readonly} removes contenteditable property to thought cannot be selected/navigated
+
+    setEditingValue(newValue)
+
+    if (newValue === oldValue) {
+      if (readonly || uneditable || options) {
+        error(null)
+        setContentInvalidState(false)
+        setInvalidState(false)
+      }
+      return
+    }
+
+    const oldValueClean = oldValue === EM_TOKEN ? 'em' : ellipsize(oldValue)
+    if (readonly) {
+      error(`"${ellipsize(oldValueClean)}" is read-only and cannot be edited.`)
+      return
+    }
+    else if (uneditable) {
+      error(`"${ellipsize(oldValueClean)}" is uneditable.`)
+      return
+    }
+    else if (options && !options.includes(newValue.toLowerCase())) {
+      error(`Invalid Value: "${newValue}"`)
+      setContentInvalidState(true)
+      setInvalidState(true)
+      return
+    }
+
+    const newNumContext = getContexts(newValue).length
+    const isNewValueURL = isURL(newValue)
+    const contextLengthChange = newNumContext > 0 || newNumContext !== getContexts(oldValueRef.current).length
+    const urlChange = isNewValueURL || isNewValueURL !== isURL(oldValueRef.current)
```

Comment:
> I was able to remove two refs by deriving the URL and numContexts from `oldValueRef` in [12d0f6f](https://github.com/cybersemics/em/commit/12d0f6f045bf9c5135dbc34186f2f4c56bda6271). Seems cleaner to me. What do you think?

That's great! It better and much cleaner.

Reply:
Hey Raine! I just found one minor mistake in the logic. With the change `thoughtChangeHandler` is being called immediately. To fix this we should deccrease the `oldValue` context length by 1.

```
const contextLengthChange = newNumContext > 0 || newNumContext !== getContexts(oldValueRef.current).length - 1
```

It is because after `existingThoughtChange` context length for the `oldValue` will be increased by 1.